### DISCUSSION TITLE
appium 2.0: driver install

### DISCUSSION THREAD

**Discussion by RSM-SBA** [appium/appium#15812] [General]
I do install appium via npm but not globally.
I have noticed that installing the driver with "node . driver install uiautomator2"
i get this error.

✖ Installing 'uiautomator2' using NPM install spec 'appium-uiautomator2-driver'
Error: Encountered an error when installing package: Command 'npm install --no-save --no-package-lock appium-uiautomator2-driver -json' exited with code 1

seem like i have to use "sudo" with that command. Is that really necessary ? This could be a problem on a jenkins remote system.

And is it possible to install the needed driver directly with the "npm install appium" command ?

Also i start the server programmatically from java and i get:
 [Appium] Available plugins:
2021-09-06 15:01:28:413 - [Appium]   - images@1.2.0
2021-09-06 15:01:28:413 - [Appium] No plugins activated. Use the --plugins flag with names of plugins to activate

is it possible to activate the plugins also with the AppiumServiceBuilder ?

**Comment by KazuCocoa** [comment]
How did you install appium?

The below command worked.
```
$ mkdir tmp
$ cd tmp
$ npm install appium@next
$ node_modules/.bin/appium  driver install uiautomator2
```
Please make sure `~/.appium` owner is a proper user since Appium 2.0 installs npm packaged in `~/.appium`, which can customize with `--appium-home` argument.

**Reply by KazuCocoa** [reply]
If your `~/.appium` had root permission, I guess you installed appium 2.0 plugins/drivers as root permission before.


For Java command:
I don't think Java clients support appium 2.0 command arguments well. Once these arguments fix as appium 2.0, anyone can add them in Java-client.

**Comment by RSM-SBA** [comment]
@KazuCocoa 
Thank you for your help,the folder ~/.appium has full access rights for all users, and the driver installation has installed files there, but still i get the error code 1. I have another look when i reinstall next time.

For the AppiumServerBuilder i found out that i can send serverArguments on creation. That seem to work.
`builder = new AppiumServiceBuilder().withLogFile(new File("appium.log")).withArgument(() -> "--base-path", "/wd/hub").withArgument(() -> "--plugins", "images");`

But now, regarless if i start the server manuall or via AppiumServerBuilder
i get an error on driver initialization

` The session identified by 437e3711-ee21-41eb-bb9d-dcc72e95cf2c is not known`

That happens on calling
`driver = new AndroidDriver<MobileElement>(new URL("http://127.0.0.1:4726/wd/hub"), caps);
`

Anyone an Idea ? or had a similar issue ?

**Reply by RSM-SBA** [reply]
ok the solution was simple....
i had an old uiautomater on the emulator...
Deleting it solved the problem with the session not found.

**Reply by RSM-SBA** [reply]
@KazuCocoa 
Question regarding the old uiautomator server on the emulator.
Should the driver initialziation delete lower versions of the uiautomator server on the device instead trying to connect ?

**Reply by KazuCocoa** [reply]
mm, usually if the installed one was older than the new one, then https://github.com/appium/appium-uiautomator2-driver/blob/3c5a0381a17f019c85fdcc79b4f688f40828cd3e/lib/uiautomator2.js#L97-L158 deletes the old one.

**Reply by RSM-SBA** [reply]
i haven't retried that problem, but at my first appium 2.0 setup, my simulator had still the uiautomator server from appium 1.2x installed, which leaded to the error 
`The session identified by 437e3711-ee21-41eb-bb9d-dcc72e95cf2c is not known`
After removing it manually from the emulator, the error was gone.

**Comment by RSM-SBA** [comment]
@KazuCocoa 
I still have the problem with the exit code 1 on installation on a jenkins machine.

I tried to create the .appium folder on a custom location and adding full access rights.
Looks like the installer writing in this folders
![image](https://user-images.githubusercontent.com/90029825/132472271-fb79096b-d7ae-4089-838c-f93d42b68023.png)

I have no idea if the installer is trying to writing to other locations also

LOG: 
mkdir /Users/jenkins/HomeOfQA/workspace/RMA-Android-Smoke-Java/uitesting/.appium
chmod 777 /Users/jenkins/HomeOfQA/workspace/RMA-Android-Smoke-Java/uitesting/.appium
npm install appium@next --drivers=uiautomator2 --plugins=images --appium-home=/Users/jenkins/HomeOfQA/workspace/RMA-Android-Smoke-Java/uitesting/.appium
appium@2.0.0-beta.13 postinstall /Users/jenkins/HomeOfQA/workspace/RMA-Android-Smoke-Java/uitesting/node_modules/appium
node ./postinstall.js

Installing the driver uiautomator2...
There was an error checking and installing driver uiautomator2: Encountered an error when installing package: Command 'npm install --no-save --no-package-lock appium-uiautomator2-driver -json' exited with code 1
Installing the plugin images...
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN uitesting No description

 appium@2.0.0-beta.13
added 425 packages from 518 contributors and audited 425 packages in 46.773s`

**Reply by KazuCocoa** [reply]
I guess something more general `npm` and git combination stuff, so run `npm install` command in git managed area...

What about 

```
$ npm install appium@next
$ node_modules/.bin/appium driver install uiautomator2 --appium-home=/Users/jenkins/HomeOfQA/workspace/RMA-Android-Smoke-Java/uitesting/.appium
$ node_modules/.bin/appium plugin install images --appium-home=/Users/jenkins/HomeOfQA/workspace/RMA-Android-Smoke-Java/uitesting/.appium
```

**Reply by RSM-SBA** [reply]
Thank you for your tips @KazuCocoa 
I have figure out that the installer could not load the chromedriver (maybe company proxy problem).
With the flag --chromedriver-skip-install the installation works.

`npm install appium@next --chromedriver-skip-install --drivers=uiautomator2 --plugins=images --appium-home=/Users/jenkins/HomeOfQA/workspace/RMA-Android-Smoke-Java/uitesting/.appium`

**Reply by KazuCocoa** [reply]
oo... i see. thanks for the info.

**Reply by jlipps** [reply]
the error message could probably be improved to show the NPM error output which might have made this an easier problem to diagnose.

**Comment by RSM-SBA** [comment]
@jlipps 
Is the chromedriver part of the uiautomator2 driver ?
In the past it was possible to add the --chromedriver-skip-install flag to the npm installation.
This is currently also possible when i install appium with 
`npm install appium@next --chromedriver-skip-install --drivers=uiautomator2`

But its not possible if i want to install via
`appium driver install uiautomator2`

Would it be possible to add the flag there also ?
If yes, i would create a feature request.

**Reply by KazuCocoa** [reply]
Probably yes, it is helpful to create an issue.

For now, you also can set the environment variable as `APPIUM_SKIP_CHROMEDRIVER_INSTALL=true appium driver install uiautomator2`
https://github.com/appium/appium-chromedriver#skipping-binary-installation
